### ISSUE TITLE
RTDN and Firebase

### ISSUE THREAD

**Issue by wdavies973** [android/play-billing-samples#342]
I am trying to manage play billing subscriptions using Firebase custom claims.

Essentially, without going into too much detail, Firebase would have a custom claim `paid` that is `true` or `false` indicating whether a user is subscribed. This claim needs to be manually updated via the Firebase SDK.

First, obviously, when a user immediately purchases a subscription, I call a Firebase cloud function that will validate the user's purchase and set `paid` to `true` when the purchase token is validated using the Play Billing SDK.

Then, my Firebase functions also subscribe to the RTDN for the subscription, and will update the `paid` claim depending on the notification. I have several questions regarding this process.

Note: I define a "claim check" as a function that takes a `purchaseToken` as an argument, validates the purchase (and determines whether the subscription is active or not), and updates the `paid` claim accordingly.

- How reliable are RTDN's? Do I have to acknowledge that I've received and processed them? For example, let's say the user cancels, and then the user renews, but the `SUBSCRIPTION_RENEWED` notification is dropped. Or maybe a `NOTIFICATION_CANCELED` is dropped. This is obviously not good. I guess the best alternative is to force a `paid` claim update every time the user opens the app and `paid` is `false`? Is there a way to acknowledge RTDN's? Do they have a retry policy?
- As RTDN's are asynchronous (relative to an Android user subscribing), I would obviously need a synchronous call from the Android app to force a claim check (it probably isn't wise to poll until a RTDN to comes in). The issue with this is then I'd be making two claim checks (one for the RTDN and one when the Android user subscribes). This isn't that big of a deal, but doesn't feel clean. Is there a way to avoid this?
- With this architecture, do I need to store `purchaseToken`s or `linkedPurchaseToken`s? It seems like if I react to subscription changes immediately, and update `paid` accordingly, I have no need for a `purchaseToken` (as the most recent one will be available to me immediately). I guess I read something about storing old `purchaseToken`s to make sure they aren't used again? But otherwise, it seems like I'd only need to store them for logging purposes?

Thanks for the help!

**Comment by conde2** [comment]
I have the same problem, need to update user claism according to the RTDN answers.

Have you managed to do it and can share with us ? @widavies

**Comment by widavies** [comment]
@conde2 I actually ended up using [RevenueCat](https://www.revenuecat.com/), it handles everything for you and is pretty easy to integrate. A lot less headache then having to completely design a subscription backend.

**Comment by conde2** [comment]
@widavies the subscription backend is already design in this repository https://github.com/android/play-billing-samples/tree/main/ClassyTaxiServer

I will be looking at this RevenueCat but still I will need to setup user claims to my users, is it possible with RevenueCat?

**Comment by widavies** [comment]
@conde2 Yeah, RevenueCat maintains a subscribed status for users. RevenueCat basically just does what `ClassyTaxiServer` does, but makes it easier and so you don't have to run your own server.
